### PR TITLE
bitmap should optimize for sparse case

### PR DESCRIPTION
**Pull request by jensenojs**
**What type of PR is this?**

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

**Which issue(s) this PR fixes:**

issue #4031

**What this PR does / why we need it:**

Not Available

**Special notes for your reviewer:**

Not Available

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

File: pkg/common/bitmap/bitmap.go [hunk 1]
```diff
@@ -47,11 +76,51 @@ func (n *Bitmap) Iterator() Iterator {
...
+
 func (itr *BitmapIterator) HasNext() bool {
-	for ; itr.i < uint64(itr.bm.len); itr.i++ {
-		if itr.bm.Contains(itr.i) {
+	// if the uint64 is 0, move forward to next word
+	// if the uint64 is not 0, then calculate the rightest_one position in a word and add it to itr.i
+	// either case loop over words not bits.
+	nwords := (itr.bm.len-1)/64 + 1
+	current_word := wordnum(itr.i)
+	mask := ((bitmask)(0) | 0xFFFFFFFF) << bitnum(itr.i) // ignore bits before check before
```
Comment (fengttt) [inline pkg/common/bitmap/bitmap.go:111]: Are you using uint32 or uint64?    I think this 0xFFF... does not have enough F in it.

File: pkg/common/bitmap/bitmap.go [hunk 2]
```diff
@@ -47,11 +76,51 @@ func (n *Bitmap) Iterator() Iterator {
 	return &BitmapIterator{i: 0, bm: n}
 }
 
+func wordnum(i uint64) uint64 {
+	return i / bits_per_bitmapword
```
Comment (fengttt) [inline pkg/common/bitmap/bitmap.go:80]: I don't know if golang compiler actually will turn this into shift and mask, but if we are doing bits, you should shift this.   Actually easier to understand in the context of bit mask.

File: pkg/common/bitmap/bitmap.go [hunk 3]
```diff
@@ -47,11 +76,51 @@ func (n *Bitmap) Iterator() Iterator {
...
+		word := itr.bm.data[current_word]
+		word &= mask
+
+		if word != 0 {
+			result := rightmost_one_pos_64(word)
+			itr.i += result
 			return true
 		}
+		itr.i += 64
+		mask = ((bitmask)(0) | 0xFFFFFFFF) // in subsequent words, consider all bits
```
Comment (fengttt) [inline pkg/common/bitmap/bitmap.go:123]: here as well, 64 bits?

**Review by fengttt — CHANGES_REQUESTED** [review]

File: pkg/common/bitmap/bitmap_test.go [hunk 4]
```diff
@@ -121,3 +121,25 @@ func TestC(t *testing.T) {
...
+
+	itr := np.Iterator()
+	i := uint64(0)
+	for itr.HasNext() {
+		r := itr.Next()
+		if (r < uint64(63)){
+			require.Equal(t, i, r)
+			i++ 
+		}else{
+			fmt.Println("i now is ", i)
```
Comment (lni) [inline pkg/common/bitmap/bitmap_test.go:141]: use T.logf() instead of Println

Comment (jensenojs) [reply pkg/common/bitmap/bitmap_test.go:141]: sorry and ok, i haven't finished it yet, just wanted to go home and write the rest part.

**Review by nnsgmsone — APPROVED** [review]

File: pkg/common/bitmap/bitmap.go (hunk 1 above)
Comment (jensenojs) [reply pkg/common/bitmap/bitmap.go]: A better implementation has been used

**Review by fengttt — APPROVED** [review]

**Review by lni — APPROVED** [review]

**Review by yingfeng — APPROVED** [review]

**Pull request merged**